tools/agent-ctl/src/client.rs

```diff
@@ -294,25 +295,135 @@ fn client_create_vsock_fd(cid: libc::c_uint, port: u32) -> Result<RawFd> {
     Ok(fd)
 }
 
-fn create_ttrpc_client(cid: libc::c_uint, port: u32) -> Result<ttrpc::Client> {
-    let fd = client_create_vsock_fd(cid, port).map_err(|e| {
-        anyhow!(format!(
-            "failed to create VSOCK connection (check agent is running): {:?}",
-            e
-        ))
-    })?;
+fn create_ttrpc_client(server_address: String) -> Result<ttrpc::Client> {
+    if server_address == "" {
+        return Err(anyhow!("server address cannot be blank"));
+    }
+
+    let fields: Vec<&str> = server_address.split("://").collect();
+
+    if fields.len() != 2 {
+        return Err(anyhow!("invalid server address URI"));
+    }
+
+    let scheme = fields[0].to_lowercase();
+
+    let fd: RawFd = match scheme.as_str() {
+        // Formats:
+        //
+        // - "unix://absolute-path" (domain socket)
+        //   (example: "unix:///tmp/domain.socket")
+        //
+        // - "unix://@absolute-path" (abstract socket)
+        //   (example: "unix://@/tmp/abstract.socket")
+        //
+        "unix" => {
+            let mut abstract_socket = false;
+
+            let mut path = fields[1].to_string();
+
+            if path.starts_with('@') {
+                abstract_socket = true;
+
+                // Remove the magic abstract-socket request character ('@')
+                // and crucially add a trailing nul terminator (required to
+                // interoperate with the ttrpc crate).
+                path = path[1..].to_string() + &"\x00".to_string();
+            }
+
+            if abstract_socket {
+                let socket_fd = match socket(
+                    AddressFamily::Unix,
+                    SockType::Stream,
+                    SockFlag::empty(),
+                    None,
+                ) {
+                    Ok(s) => s,
+                    Err(e) => return Err(anyhow!("Failed to create Unix Domain socket: {:?}", e)),
```

Comment:
Q, why don't use `context`?  

```rust
anyhow!(e).context("Failed to create Unix Domain socket"),
```

Reply:
yes, the  context/with_context can make it much neat.

Reply:
Nice! Fixed!